File: 백준/Silver/15656. N과 M （7）/N과 M （7）.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class Main {
	public static void main(String[] args) throws IOException {
		input();
		solution();
		output();
	}
	private static int N, M;
	private static int[] arr;
	private static List<String> list;
	private static void solution() {
		list = new ArrayList<>();
		int[] res = new int[M];
		perm(res, 0);
	}
	private static void perm(int[] res, int depth) {
		if (depth == M) {
			list.add(IntStream.of(res).mapToObj(String::valueOf).collect(Collectors.joining(" ")));
			return;
		}

		for (int i = 0; i < N; i++) {
			res[depth] = arr[i];
			perm(res, depth + 1);
		}
	}
	private static void output() {
		System.out.println(String.join("\n", list));
	}
	private static void input() throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int[] nm = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
		arr = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).sorted().toArray();
		br.close();
		N = nm[0];
		M = nm[1];
	}
}

File: 백준/Silver/15656. N과 M （7）/N과 M （7）.kt
fun main() = with(System.`in`.bufferedReader()) {
    val (n, m) = readln().split(" ").map { it.toInt() }
    val arr = readln().split(" ").map { it.toInt() }.sorted().toIntArray()
    close()

    val tmp = IntArray(m)
    perm(arr, tmp, 0, m)
    println(sb.toString())
}

private fun perm(arr : IntArray, tmp : IntArray, depth : Int, m : Int) {
    if (depth == m) {
        toPrint(tmp)
        return
    }

    for (i in arr.indices) {
        tmp[depth] = arr[i]
        perm(arr, tmp, depth + 1, m)
    }
}

private val sb = StringBuilder()

private fun toPrint(arr : IntArray) =
    sb.append(
        arr.joinToString(" ")
    ).append("\n")